### ISSUE TITLE
C# compiler should not report CS1983 when ValueTask<T> is returned from async method.

### ISSUE THREAD

**Issue by TaffarelJr** [dotnet/roslyn#24678]
This seems to be the same issue as #19616, and I read through #17417 and #16530 as well. But since they're all closed, and it seems like it's been a while since #12621 was last worked on, I created this new issue.

I've been working with `ValueTask<T>` over the past month or so, and it's been OK (with the `SuppressMessage` attribute), until just today I got this error that I can't get rid of. I wonder if this issue could have re-surfaced in a recent patch or something.

**Version Used**: 
Visual Studio 15.5.6

**Steps to Reproduce**:
Simple solution containing one `.csproj` file:
```
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFramework>netstandard1.0</TargetFramework>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.0.0" />
    </ItemGroup>

</Project>
```

and one class:
```
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;

#pragma warning disable CS1983 // The return type of an async method must be void, Task or Task<T>

namespace ClassLibrary
{
    public class MyClass
    {
        [SuppressMessage("Concurrency", "CS1983", Justification = "Returns a ValueTask instead of a Task, but is still awaitable.")]
        public async ValueTask<int> DoStuffAsync()
        {
            await Task.Delay(100);
            return 5;
        }
    }
}
```

**Expected Behavior**:
This code is just a simplistic example, but it should still compile without errors or warnings.

**Actual Behavior**:
Visual Studio shows error `CS1983: The return type of an async method must be void, Task or Task<T>` (with no context help to fix it) until I add either the `SuppressMessage` attribute or the `#pragma`. But even with both of those in place, when I go to build, the C# compiler still throws the same exception:

```
1>  Task "Csc"
1>    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe /noconfig /unsafe- /checked- /nowarn:1701,1702,1705,2008 /nostdlib+ /errorreport:prompt /warn:4 /define:TRACE;DEBUG;NETSTANDARD1_0 /errorendlocation /preferreduilang:en-US /reference:C:\Users\...\.nuget\packages\system.collections\4.3.0\ref\netstandard1.0\System.Collections.dll /reference:C:\Users\...\.nuget\packages\system.diagnostics.debug\4.3.0\ref\netstandard1.0\System.Diagnostics.Debug.dll /reference:C:\Users\...\.nuget\packages\system.diagnostics.tools\4.3.0\ref\netstandard1.0\System.Diagnostics.Tools.dll /reference:C:\Users\...\.nuget\packages\system.globalization\4.3.0\ref\netstandard1.0\System.Globalization.dll /reference:C:\Users\...\.nuget\packages\system.io\4.3.0\ref\netstandard1.0\System.IO.dll /reference:C:\Users\...\.nuget\packages\system.linq\4.3.0\ref\netstandard1.0\System.Linq.dll /reference:C:\Users\...\.nuget\packages\system.linq.expressions\4.3.0\ref\netstandard1.0\System.Linq.Expressions.dll /reference:C:\Users\...\.nuget\packages\system.net.primitives\4.3.0\ref\netstandard1.0\System.Net.Primitives.dll /reference:C:\Users\...\.nuget\packages\system.objectmodel\4.3.0\ref\netstandard1.0\System.ObjectModel.dll /reference:C:\Users\...\.nuget\packages\system.reflection\4.3.0\ref\netstandard1.0\System.Reflection.dll /reference:C:\Users\...\.nuget\packages\system.reflection.extensions\4.3.0\ref\netstandard1.0\System.Reflection.Extensions.dll /reference:C:\Users\...\.nuget\packages\system.reflection.primitives\4.3.0\ref\netstandard1.0\System.Reflection.Primitives.dll /reference:C:\Users\...\.nuget\packages\system.resources.resourcemanager\4.3.0\ref\netstandard1.0\System.Resources.ResourceManager.dll /reference:C:\Users\...\.nuget\packages\system.runtime\4.3.0\ref\netstandard1.0\System.Runtime.dll /reference:C:\Users\...\.nuget\packages\system.runtime.extensions\4.3.0\ref\netstandard1.0\System.Runtime.Extensions.dll /reference:C:\Users\...\.nuget\packages\system.text.encoding\4.3.0\ref\netstandard1.0\System.Text.Encoding.dll /reference:C:\Users\...\.nuget\packages\system.text.encoding.extensions\4.3.0\ref\netstandard1.0\System.Text.Encoding.Extensions.dll /reference:C:\Users\...\.nuget\packages\system.text.regularexpressions\4.3.0\ref\netstandard1.0\System.Text.RegularExpressions.dll /reference:C:\Users\...\.nuget\packages\system.threading\4.3.0\ref\netstandard1.0\System.Threading.dll /reference:C:\Users\...\.nuget\packages\system.threading.tasks\4.3.0\ref\netstandard1.0\System.Threading.Tasks.dll /reference:C:\Users\...\.nuget\packages\system.threading.tasks.extensions\4.0.0\lib\netstandard1.0\System.Threading.Tasks.Extensions.dll /reference:C:\Users\...\.nuget\packages\system.xml.readerwriter\4.3.0\ref\netstandard1.0\System.Xml.ReaderWriter.dll /reference:C:\Users\...\.nuget\packages\system.xml.xdocument\4.3.0\ref\netstandard1.0\System.Xml.XDocument.dll /debug+ /debug:portable /filealign:512 /nologo /optimize- /out:obj\Debug\netstandard1.0\ClassLibrary.dll /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" /target:library /warnaserror- /utf8output /deterministic+ MyClass.cs "C:\Users\...\AppData\Local\Temp\.NETStandard,Version=v1.0.AssemblyAttributes.cs" obj\Debug\netstandard1.0\\TemporaryGeneratedFile_E7A71F73-0F8D-4B9B-B56E-8E70B10BC5D3.cs obj\Debug\netstandard1.0\\TemporaryGeneratedFile_036C0B5B-1481-4323-8D20-8F5ADCB23D92.cs obj\Debug\netstandard1.0\\TemporaryGeneratedFile_5937a670-0e60-4077-877b-f7221da3dda1.cs obj\Debug\netstandard1.0\ClassLibrary.AssemblyInfo.cs /warnaserror+:NU1605
1>    Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn
1>    MyClass.cs(11,31,11,43): error CS1983: The return type of an async method must be void, Task or Task<T>
1>  Done executing task "Csc" -- FAILED.
```

**Comment by jcouv** [comment]
Tagging @cston

**Comment by cston** [comment]
@TaffarelJr Thanks for reporting this issue.

It looks like `ValueTask<T>` from `System.Threading.Tasks.Extensions, Version=4.0.0.0` does not have the `[AsyncMethodBuilder]` attribute that the compiler uses to rewrite calls to the `async` method. The is included on `ValueTask<T>` in `Version=4.1.0.0`. Are you able to move to a more recent version of the assembly?

**Comment by TaffarelJr** [comment]
Ahh, cripes. That was it. I changed my NuGet reference from `4.0.0` to `4.3.0`, and it looks like the messages went away. Thanks so much, you guys are awesome!

**Comment by jcouv** [comment]
Good to hear. I'll go ahead and close the issue.
